pkg/controller/util/propagatedversion.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"sort"
+	"strconv"
+
+	fedv1a1 "github.com/marun/federation-v2/pkg/apis/federation/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+)
+
+type VersionCompareType int
+
+const (
+	ResourceVersion VersionCompareType = iota
+	Generation
+)
+
+type ComparisonHelper interface {
+	GetVersion(objectMeta *metav1.ObjectMeta) string
+	Equivalent(objMeta1, objectMeta2 *metav1.ObjectMeta) bool
+}
+
+// NewComparisonHelper instantiates and returns a Resource or Generation Helper
+// struct that implements the ComparisonHelper interface based on the version
+// comparison type passed in.
+func NewComparisonHelper(compareType VersionCompareType) (ComparisonHelper, error) {
+	switch compareType {
+	case ResourceVersion:
+		return &ResourceHelper{}, nil
+	case Generation:
+		return &GenerationHelper{}, nil
+	default:
+		err := fmt.Errorf("Unrecognized version comparison type %v", compareType)
+		runtime.HandleError(err)
```

Comment:
Use of `HandleError` should be restricted to functions that don't return an error.  If an error is returned, the caller can decide what to do with it.